[dynamo] Add support for DELETE_SUBSCR

arunppsg opened the pull request:
Fixes #123317


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

arunppsg committed:
add support for DELETE_SUBSCR

jansel commented:
Can you add a test?

arunppsg committed:
added test

jansel commented:
@pytorchbot merge

arunppsg commented:
@jansel the failing tests passes locally in machine. I am not sure how to fix it. Could you suggest me directions for fixing it?

jansel commented:
@arunppsg you need to run the failing tests with `PYTORCH_TEST_WITH_DYNAMO=1`

For example:
```
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_reductions.py -k test_warn_invalid_degrees_of_freedom_cpu_complex64
```

These are regular (eager) pytorch tests run with TorchDynamo.  In some cases, the failures will be expected and your PR might be uncovering unrelated issues.  If that is the case you can run:
```
./scripts/compile_tests/update_failures.py <sha256_hash_of_commit>
````

Which will mark any new failures as expected.

arunppsg commented:
I think the failures are unrelated. I replaced this line `self.items.__delitem__(Hashable(args[0]))` with a call to `self.items.pop(Hashable(args[0]))` (using pop instead of __delitem__) and the test still fails. Could you confirm that the failures are unrelated, so that we could get this merged in?

jansel commented:
@arunppsg if the failures are unrelated you need to run:
```
./scripts/compile_tests/update_failures.py 62907be5c1c20081bc47237c4d6375a4c45e5c66
```
Then commit the updates to expected failures that generates.

This is causing new failures since we are no longer falling back to eager in some tests.

arunppsg committed:
added expected failures

jansel commented:
@pytorchbot merge

arunppsg committed:
remove test successes

arunppsg commented:
rebased with main and removed test which resulted in successes in [0bb32c5](https://github.com/pytorch/pytorch/pull/123526/commits/0bb32c5ae3018b9b42bd5b9f4ea35c35f93d6567)

jansel commented:
@pytorchbot merge

jansel committed:
Update test failures

jansel commented:
I updated the test failure for you

@pytorchbot merge

arunppsg committed:
update test failures

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot rebase

jansel commented:
@pytorchbot merge

jansel committed:
fix infinite recursion in sys.modules handling

jansel commented:
@pytorchbot merge

arunppsg commented:
Thanks @jansel
